File: apps/mobile/src/modules/photo-viewer/photoInfoGear.ts
import type { GalleryExif, GalleryPhoto } from '@/modules/galleries/types'

import type { Translator } from './photoInfoFormatters'
import {
  formatApertureGlyph,
  formatDimensions,
  formatExposureBias,
  formatExposureTime,
  formatFileSize,
  formatFilmMode,
  formatFocalPair,
  formatIso,
  formatMegapixels,
  formatToneType,
  joinMakeAndModel,
  textValue,
} from './photoInfoFormatters'

export interface PhotoInfoGear {
  model: string
  formatBadge: string | null
  styleBadge: string | null
  lens: string | null
  rating: number
  specs: string[]
  tone: string | null
  exposure: string[]
}

function compact(values: (string | null)[]): string[] {
  return values.filter((value): value is string => value !== null)
}

export function buildPhotoInfoGear(
  photo: GalleryPhoto,
  exif: GalleryExif | null,
  t: Translator,
  locale: string,
): PhotoInfoGear {
  const lens = joinMakeAndModel(exif?.LensMake, exif?.LensModel) ?? textValue(photo.lens)
  const model = joinMakeAndModel(exif?.Make, exif?.Model) ?? textValue(photo.camera) ?? textValue(photo.title) ?? ''

  return {
    model,
    formatBadge: textValue(photo.format)?.toUpperCase() ?? null,
    styleBadge: formatFilmMode(exif?.FujiRecipe?.FilmMode),
    lens,
    rating: Math.min(5, Math.max(0, Math.round(photo.rating ?? 0))),
    specs: compact([
      formatMegapixels(photo.width, photo.height, locale),
      formatDimensions(photo.width, photo.height),
      formatFileSize(photo.size, locale),
    ]),
    tone: formatToneType(photo.toneAnalysis, t),
    exposure: compact([
      formatIso(exif?.ISO),
      formatFocalPair(exif?.FocalLength, exif?.FocalLengthIn35mmFormat),
      formatExposureBias(exif?.ExposureCompensation, locale),
      formatApertureGlyph(exif?.FNumber),
      formatExposureTime(exif?.ExposureTime ?? exif?.ShutterSpeedValue ?? exif?.ShutterSpeed, locale),
    ]),
  }
}

File: apps/mobile/modules/photo-masonry/ios/Models/PhotoInfoGear.swift
import Foundation

struct PhotoInfoGear: Codable, Equatable, Sendable {
  let model: String
  let formatBadge: String?
  let styleBadge: String?
  let lens: String?
  let rating: Int
  let specs: [String]
  let tone: String?
  let exposure: [String]

  static func build(
    photo: GalleryPhoto,
    exif: GalleryExif?,
    localeIdentifier: String
  ) -> PhotoInfoGear {
    let lens = PhotoInfoFormatters.joinMakeAndModel(exif?["LensMake"], exif?["LensModel"])
      ?? PhotoInfoFormatters.text(photo.lens)
    let model = PhotoInfoFormatters.joinMakeAndModel(exif?["Make"], exif?["Model"])
      ?? PhotoInfoFormatters.text(photo.camera)
      ?? PhotoInfoFormatters.text(photo.title)
      ?? ""
    return PhotoInfoGear(
      model: model,
      formatBadge: PhotoInfoFormatters.text(photo.format)?.uppercased(),
      styleBadge: PhotoInfoFormatters.formatFilmMode(exif?["FujiRecipe"]?.object?["FilmMode"]),
      lens: lens,
      rating: min(5, max(0, photo.rating ?? 0)),
      specs: [
        PhotoInfoFormatters.formatMegapixels(photo.width, photo.height, localeIdentifier: localeIdentifier),
        PhotoInfoFormatters.formatDimensions(photo.width, photo.height),
        PhotoInfoFormatters.formatFileSize(photo.size, localeIdentifier: localeIdentifier),
      ].compactMap { $0 },
      tone: PhotoInfoFormatters.formatToneType(photo.toneAnalysis),
      exposure: [
        PhotoInfoFormatters.formatISO(exif?["ISO"]),
        PhotoInfoFormatters.formatFocalPair(exif?["FocalLength"], exif?["FocalLengthIn35mmFormat"]),
        PhotoInfoFormatters.formatExposureBias(exif?["ExposureCompensation"], localeIdentifier: localeIdentifier),
        PhotoInfoFormatters.formatApertureGlyph(exif?["FNumber"]),
        PhotoInfoFormatters.formatExposureTime(
          exif?["ExposureTime"] ?? exif?["ShutterSpeedValue"] ?? exif?["ShutterSpeed"],
          localeIdentifier: localeIdentifier
        ),
      ].compactMap { $0 }
    )
  }
}
